controllers/topology/reconcile_state_test.go

```diff
@@ -0,0 +1,361 @@
+	clusterInfrastructure3 := newFakeInfrastructureCluster(metav1.NamespaceDefault, "infrastructure-cluster3").Obj()
+	clusterInfrastructure3WithInstanceSpecificChanges := clusterInfrastructure3.DeepCopy()
+	clusterInfrastructure3WithInstanceSpecificChanges.SetLabels(map[string]string{"foo": "bar"})
+	clusterInfrastructure4 := newFakeInfrastructureCluster(metav1.NamespaceDefault, "infrastructure-cluster4").Obj()
+	clusterInfrastructure4WithTemplateOverridingChanges := clusterInfrastructure4.DeepCopy()
+	err := unstructured.SetNestedField(clusterInfrastructure4WithTemplateOverridingChanges.UnstructuredContent(), false, "spec", "fakeSetting")
+	g.Expect(err).ToNot(HaveOccurred())
+	clusterInfrastructure5 := newFakeInfrastructureCluster(metav1.NamespaceDefault, "infrastructure-cluster5").Obj()
+
+	tests := []struct {
+		name    string
+		current *unstructured.Unstructured
+		desired *unstructured.Unstructured
+		want    *unstructured.Unstructured
+		wantErr bool
+	}{
+		{
+			name:    "Should create desired InfrastructureCluster if the current does not exists yet",
+			current: nil,
+			desired: clusterInfrastructure1,
+			want:    clusterInfrastructure1,
+			wantErr: false,
+		},
+		{
+			name:    "No-op if current InfrastructureCluster is equal to desired",
+			current: clusterInfrastructure2,
+			desired: clusterInfrastructure2,
+			want:    clusterInfrastructure2,
+			wantErr: false,
+		},
+		{
+			name:    "Should preserve instance specific changes",
+			current: clusterInfrastructure3WithInstanceSpecificChanges,
+			desired: clusterInfrastructure3,
+			want:    clusterInfrastructure3WithInstanceSpecificChanges,
+			wantErr: false,
+		},
+		{
+			name:    "Should restore template values if overridden",
+			current: clusterInfrastructure4WithTemplateOverridingChanges,
+			desired: clusterInfrastructure4,
+			want:    clusterInfrastructure4,
+			wantErr: false,
+		},
+		{
+			name:    "Fails for incompatible changes",
+			current: clusterInfrastructure5,
+			desired: clusterInfrastructure1,
+			wantErr: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			g := NewWithT(t)
+
+			fakeObjs := make([]client.Object, 0)
+			if tt.current != nil {
+				fakeObjs = append(fakeObjs, tt.current)
+			}
+			fakeClient := fake.NewClientBuilder().
+				WithScheme(fakeScheme).
+				WithObjects(fakeObjs...).
+				Build()
+
+			currentState := &clusterTopologyState{infrastructureCluster: tt.current}
+
+			// TODO: stop setting ResourceVersion when building objects
+			tt.desired.SetResourceVersion("")
+			desiredState := &clusterTopologyState{infrastructureCluster: tt.desired}
+
+			r := ClusterReconciler{
+				Client: fakeClient,
+			}
+			err := r.reconcileInfrastructureCluster(ctx, currentState, desiredState)
+			if tt.wantErr {
+				g.Expect(err).To(HaveOccurred())
+				return
+			}
+			g.Expect(err).ToNot(HaveOccurred())
+
+			got := tt.want.DeepCopy()
```

Comment:
I think it would be good to add a comment to mention that we have to DeepCopy want for TypeMeta so that the get works. I was confused at first.